apps/rush-lib/src/logic/operations/OperationExecutionRecord.ts

```diff
@@ -0,0 +1,113 @@
+import { Operation } from './Operation';
+import { Stopwatch } from '../../utilities/Stopwatch';
+
+export interface IOperationExecutionRecordContext {
+  streamCollator: StreamCollator;
+
+  debugMode: boolean;
+  quietMode: boolean;
+}
+
+/**
+ * Internal class representing everything about executing an operation
+ */
+export class OperationExecutionRecord implements IOperationRunnerContext {
+  /**
+   * The current execution status of an operation. Operations start in the 'ready' state,
+   * but can be 'blocked' if an upstream operation failed. It is 'executing' when
+   * the operation is executing. Once execution is complete, it is either 'success' or
+   * 'failure'.
+   */
+  public status: OperationStatus = OperationStatus.Ready;
+
+  /**
+   * The error which occurred while executing this operation, this is stored in case we need
+   * it later (for example to re-print errors at end of execution).
+   */
+  public error: OperationError | undefined = undefined;
+
+  /**
+   * This number represents how far away this Operation is from the furthest "root" operation (i.e.
+   * an operation with no consumers). This helps us to calculate the critical path (i.e. the
+   * longest chain of projects which must be executed in order, thereby limiting execution speed
+   * of the entire operation tree.
+   *
+   * This number is calculated via a memoized depth-first search, and when choosing the next
+   * operation to execute, the operation with the highest criticalPathLength is chosen.
+   *
+   * Example:
+   *        (0) A
+   *             \
+   *          (1) B     C (0)         (applications)
+   *               \   /|\
+   *                \ / | \
+   *             (2) D  |  X (1)      (utilities)
+   *                    | / \
+   *                    |/   \
+   *                (2) Y     Z (2)   (other utilities)
+   *
+   * All roots (A & C) have a criticalPathLength of 0.
+   * B has a score of 1, since A depends on it.
+   * D has a score of 2, since we look at the longest chain (e.g D->B->A is longer than D->C)
+   * X has a score of 1, since the only package which depends on it is A
+   * Z has a score of 2, since only X depends on it, and X has a score of 1
+   * Y has a score of 2, since the chain Y->X->C is longer than Y->C
+   *
+   * The algorithm is implemented in AsyncOperationQueue.ts as calculateCriticalPathLength()
+   */
+  public criticalPathLength: number | undefined = undefined;
+
+  /**
+   * The set of operations that must complete before this operation executes.
+   */
+  public readonly dependencies: Set<OperationExecutionRecord> = new Set();
+  /**
+   * The set of operations that depend on this operation.
+   */
+  public readonly consumers: Set<OperationExecutionRecord> = new Set();
+
+  public readonly stopwatch: Stopwatch = new Stopwatch();
+  public readonly stdioSummarizer: StdioSummarizer = new StdioSummarizer();
+
+  public readonly runner: IOperationRunner;
+  public readonly weight: number;
```

Comment:
May need some description on how weight plays into order of execution

Reply:
The comment is in `Operation`, which is where it is a mutable field, but I can copy it here for clarity.